### PR TITLE
chore(dev): replace current flake8-based and other linting tools with `ruff`

### PR DESCRIPTION
**Pull request by cpcloud**
This PR replaces all of our flake8 checks and tools with [`ruff`](https://github.com/charliermarsh/ruff), which is Python linter implemented in Rust designed to be fast, lightweight and combine the most commonly used flake8 plugins into a single tool. While I do not care too much about the performance for ibis developement (though I do for my LSP configuration!) I do care about tool consolidation, which is the main purpose of this PR.

### CODE REVIEW

File: ibis/expr/api.py [hunk 1]
```diff
@@ -26,77 +26,13 @@
 from ibis.expr.deferred import Deferred
 from ibis.expr.schema import Schema
 from ibis.expr.sql import parse_sql, show_sql, to_sql
-from ibis.expr.types import (  # noqa: F401
-    ArrayColumn,
```
Comment (kszucs) [inline ibis/expr/api.py:30]: This looks like a breaking change. Are you sure that we should remove these?

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -25,26 +25,16 @@ repos:
...
-      - id: flake8
-        exclude: (^ibis/tests/sql/snapshots/|_py310\.py$)
+      - id: ruff
+        # makes sure to exclude a file even if it's passed in explicitly
+        args: ["--force-exclude"]
   - repo: https://github.com/pycqa/docformatter
     rev: v1.5.1
     hooks:
       - id: docformatter
-  - repo: https://github.com/MarcoGorelli/absolufy-imports
```
Comment (kszucs) [inline .pre-commit-config.yaml:37]: Is ruff handling it?

**Comment by kszucs** [comment]
@cpcloud can you add a more descriptive name to the PR? :)

File: ibis/expr/api.py (hunk 1 above)
Comment (cpcloud) [reply ibis/expr/api.py:30]: It is, but these types are stable under `ibis.expr.types as ir`. I'd rather have just one place to access them than a hodgepodge list of things that doesn't actually include everything and has to be manually maintained.

File: .pre-commit-config.yaml (hunk 2 above)
Comment (cpcloud) [reply .pre-commit-config.yaml:37]: It's handling checking for new relative imports using the [`"TID"`](https://github.com/charliermarsh/ruff#flake8-tidy-imports-tid) code.

**Comment by cpcloud** [comment]
> @cpcloud can you add a more descriptive name to the PR? :)

Done!

File: ibis/expr/api.py (hunk 1 above)
Comment (kszucs) [reply ibis/expr/api.py:30]: Can you separate this into another commit including the breaking change note?

**Review by kszucs — APPROVED** [review]
After including the breaking change note it looks good to me!

File: ibis/expr/api.py (hunk 1 above)
Comment (cpcloud) [reply ibis/expr/api.py:30]: Will do!

**Comment by cpcloud** [comment]
I'll squash into a single commit.

File: ibis/expr/api.py (hunk 1 above)
Comment (cpcloud) [reply ibis/expr/api.py:30]: Done.

**Comment by jcrist** [comment]
I have `flake8` configured with my editor in a way that I don't really want to change right now. If anyone else is in this situation, you can:

- readd the deleted `setup.cfg` to your local dev repo
- append `setup.cfg` to `.git/info/exclude` to avoid ever committing that file to the repo, without having to modify your local or global `.gitignore`.

**Pull request merged**